design/specs/list-data-type.md

```diff
@@ -0,0 +1,194 @@
+---
+title: List Data Type
+description: 
+published: false
+date: 2022-04-10 18:02:53
+tags: 
+editor: markdown
+dateCreated: 2022-04-06 18:02:58
+---
+
+# Context
+
+This document specifies the design of List data type within the Mathesar application. List data type can include one or more items that are in the same basic data type (e.g., Number), separated by a comma delimiter if more than one item is present. Mathesar aims to support all basic data types in List format. Users will be able to view data that is currently in Mathesar List data type, edit and filter List data, create new columns in List data type, and convert existing data into List data type. 
+
+The goal of the design in general is to help non-technical understand List data type and maintain List data integrity. Aggregating data into Lists via mapping tables is not beyond the scope of this spec and should be designed in visually different ways to communicate the affordance.
+
+Mathesar List data type is similar to [PostgreSQL "Arrays" concept](https://www.postgresql.org/docs/current/arrays.html). However, Mathesar List data type is only supporting single dimensional Array/List currently.
+
+Read [Design for List data type](https://github.com/centerofci/mathesar/issues/978) for a more in-depth explanation of this feature.
+
+Read [Design for visual query builder](https://github.com/centerofci/mathesar/issues/1065) for mapping table details.
+
+# Scenarios
+
+## 1. Working with Existing List Data
+
+Users will be able to view, filter, and edit existing List data. This scenario assumes no List item data type change (e.g., converting a Number item to a Text item) during the edit. 
+
+### 1.1 Viewing Existing List Data
+
+Existing Mathesar List data distinguishes itself from other data type through four unique ways: 
+
+1. Column header data type icon.
+2. Data type icon and description in the column header menu. `Allow List` option is required to be turned on for List data type.
+3. Tooltip information when hovering on the column header data type icon. 
+4. Items in a single cell is separated by comma delimiters and a `show all` option is present. Users can use `show all` function to expand the List and view all List items.
+
+---
+Wireframes
+
+[Viewing Existing List Data](https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?node-id=7317%3A80859&viewport=306%2C48%2C0.59&scaling=min-zoom&starting-point-node-id=7379%3A82626&show-proto-sidebar=1)
```

Comment:
I think it would be good to show list using pills even before the user clicks on "show all"

Reply:
I also don't think we need delimiters at all if we're showing list items as pills.

Reply:
I would advocate for clearly communicate the expected List structure to users to facilitate their editing progress. However, this is just my expectation as a single user and I'm ok with changing it to what makes sense to you.

Reply:
In this case, I believe we might also want tag-like items. Assume we're entering dates or boolean list items; if we had a control to add a new item, it would then add a tag that, when interacted with, would display a checkbox or a date picker. Also, if you're entering an email, the pill-like design could help orient the user when there are validation errors (email entered in incorrect format, missing @, etc.).

Reply:
Thanks for the clarification. I'm not confident that tags without delimiters is the best way to help users edit the List. For example, if they'd like to add two independent items by simply typing, what would be their expectation in telling Mathesar where the two items separate? Would it be a space, a comma, or something else? On the other hand, aggregated lists with an underlying mapping table wouldn't have this issue, because they are not created by typing in a cell. The attached prototype is also tag-like, just with delimiters to indicate how users should edit the List, and the interactions you mentioned should have ways to work without problems.

Reply:
I see what you mean. So in this case, when you enter a comma, does a new list item appear? how will the user insert a list item in between two existing items? How are other data types inserted into the list?